File: ccesim/device.py
import datetime as dt
import random
from ccesim.devicegroups import DeviceGroup, rtmds, fridges
from ccesim.facilities import Facility, random_facility
from ccesim.generator import random_serial, random_amid
from ccesim.simulator import SimulatedRecordSet, SimulatorState, default_config
import logging
from ccesim.schemas import (
    RtmdReport,
    EmsReport,
    TransferMetadata
)

logger = logging.getLogger(__name__)

class MonitoringDeviceConfig:
    '''Configuration class for virtual CCE monitoring devices.'''

    def __init__(self,
                 type: str =None,
                 upload_interval: int = None,
                 sample_interval: int = None,
                 facility: Facility = None,
                 manufacturer: str = None
                 ):
        # Randomly select device type if not specified
        if type is None:
            type_options = ['rtmd', 'ems']
            type_weights = [.7, .3]
            type = random.choices(type_options, weights=type_weights, k=1)[0]
        self.type = type

        # Randomly select upload and sample intervals if not specified
        if upload_interval is None:
            upload_interval_options = [3600, 7200, 14400, 28800]
            upload_interval_weights = [.5, .3, .1, .1]
            upload_interval = random.choices(upload_interval_options, weights=upload_interval_weights, k=1)[0]
        self.upload_interval = upload_interval

        # Randomly select sample interval if not specified
        if sample_interval is None:
            sample_interval_options = [600, 900]
            sample_interval_weights = [.3, .7]
            sample_interval = random.choices(sample_interval_options, weights=sample_interval_weights, k=1)[0]
        self.sample_interval = sample_interval
        if self.upload_interval < self.sample_interval:
            raise ValueError("Upload interval must be greater than or equal to sample interval.")
        if self.upload_interval % self.sample_interval != 0:
            raise ValueError("Upload interval must be a multiple of sample interval.")

        # Calculate the batch size based on the upload and sample intervals
        self.batch_size = self.upload_interval // self.sample_interval

        if isinstance(facility, Facility):
            self.facility = facility
        else:
            self.facility = random_facility()
        self.manufacturer = manufacturer

        # Generate the appliance and device based on the type
        dg_f = DeviceGroup(fridges)
        self.appliance = dg_f.random_device() # random fridge
        self.device = self.appliance   # Assume that EMS devices have monitoring devices built-in
        if self.type == 'rtmd':
            dg = DeviceGroup(rtmds)
            self.device = dg.random_device(manufacturer=self.manufacturer)


    def __repr__(self):
        return f"MonitoringDeviceConfig(type={self.type}, upload_interval={self.upload_interval}, sample_interval={self.sample_interval}, device={self.device.manufacturer} {self.device.model}, facility={self.facility.facility_name})"


def _infer_power_type(appliance_type: str) -> str:
    """Infer power type from the appliance type string in the PQS catalog.

    Appliance types containing 'Solar' or 'SDD' are solar-powered;
    everything else is mains-powered.
    """
    if appliance_type is None:
        return 'mains'
    lower = appliance_type.lower()
    if 'solar' in lower or 'sdd' in lower:
        return 'solar'
    return 'mains'


class BaseRtmDevice:
    """
    Class that models an RTMD or EMS device.

    Uses a physics-based thermal simulator to generate synthetic CCE data
    instead of replaying database records.
    """
    def __init__(self, config: MonitoringDeviceConfig):
        self.config = config
        self.last_sample_time = None  # Time of the last sample in previous report

        # Infer power source from the appliance's PQS type string
        self.powersource = _infer_power_type(self.config.appliance.type)

        # Create simulation config based on power type and facility location
        self.sim_config = default_config(
            power_type=self.powersource,
            latitude=self.config.facility.latitude,
        )
        self.sim_config.sample_interval = self.config.sample_interval

        # Simulator state (initialized on first report)
        self.simulator_state = None

        # Initialize device attributes
        self.amfr = self.config.appliance.manufacturer
        self.amod = self.config.appliance.model
        self.apqs = self.config.appliance.pqs_code
        self.aser = random_serial()
        self.adop = self._get_mock_adop()

        if self.config.type == 'rtmd':
            self.lmfr = self.config.device.manufacturer
            self.lmod = self.config.device.model
            self.lpqs = self.config.device.pqs_code
        elif self.config.type == "ems":
            self.lmfr = self.amfr
            self.lmod = self.amod
            self.lpqs = self.apqs
        else:
            raise ValueError(f"Invalid device type: {self.config.type}")

        self.lser = random_serial()
        self.ldop = self.adop
        self.lsv = '0.1.x'

        # For simplicity, use the same values for EMD and Logger
        self.emfr = self.lmfr
        self.emod = self.lmod
        self.epqs = self.lpqs
        self.eser = self.lser
        self.edop = self.ldop
        self.emsv = self.lsv

        self.cid = self.config.facility.iso
        self.lat = self.config.facility.latitude
        self.lng = self.config.facility.longitude

        # Supplier-internal appliance monitoring id, minted once and held
        # stable across all reports this device produces.
        self.amid = random_amid()

    def _build_dlst(self):
        """Build the cce-interop DLST: maps the performance properties this
        RTMD reports (TVC, TAMB) to sensor definitions.

        The sensor maker mirrors the EMD/logger manufacturer (EMFR); SIDs are
        the logger serial plus a sensor port, per the schema's $comment.
        """
        return {
            'TVC': {
                'SID': f'{self.eser}-P1',
                'SMFR': self.emfr,
                'SMOD': f'{self.emod} External Probe',
                'SDOP': self.edop,
            },
            'TAMB': {
                'SID': f'{self.eser}-P2',
                'SMFR': self.emfr,
                'SMOD': f'{self.emod} Onboard Sensor',
                'SDOP': self.edop,
            },
        }

    def __repr__(self):
        return (f"BaseRtmDevice(type={self.config.type}, "
                f"CID={self.cid}, "
                f"AMFR={self.amfr}, AMOD={self.amod}, "
                f"ASER={self.aser}, ADOP={self.adop}, "
                f"LMFR={self.lmfr}, LMOD={self.lmod}, "
                f"LSEV={self.lsv})")

    def _get_mock_adop(self):
        """Generate a random date string (YYYY-MM-DD) between Jan 1, 2018 and Dec 31, 2024."""
        start_date = dt.date(2018, 1, 1)
        end_date = dt.date(2024, 12, 31)
        random_days = random.randint(0, (end_date - start_date).days)
        random_date = start_date + dt.timedelta(days=random_days)
        return random_date.strftime('%Y-%m-%d')

    def create_report(self, report_time=None):
        '''
        Create a report for the device based on device configuration.

        This method can be called:
            1. On regular intervals (e.g. via Locust) with report_time=None (defaults to now).
            2. With a specific report_time for batch/historical data generation.

        The report covers the period from last_sample_time to report_time,
        generating enough samples to fill the gap at the configured sample_interval.
        '''

        if report_time is None:
            report_time = dt.datetime.now(dt.UTC).replace(tzinfo=None, microsecond=0)
        if not isinstance(report_time, dt.datetime):
            raise ValueError("report_time must be a datetime object.")
        self.report_time = report_time

        if self.last_sample_time is None:
            self.last_sample_time = self.report_time - dt.timedelta(seconds=self.config.upload_interval)

        if self.last_sample_time >= self.report_time:
            raise ValueError("report_time must be greater than the last sample time. Please check your report_time.")

        batch_size = (self.report_time - self.last_sample_time) // dt.timedelta(seconds=self.config.sample_interval)
        time_of_first_record_in_this_batch = self.last_sample_time + dt.timedelta(seconds=self.config.sample_interval)

        # Generate simulated records
        recordset = SimulatedRecordSet.generate(
            config=self.sim_config,
            batch_size=batch_size,
            start_time=time_of_first_record_in_this_batch,
            interval=self.config.sample_interval,
            state=self.simulator_state,
        )
        self.simulator_state = recordset.state

        nudged_coordinates = self.config.facility.get_nudged_coordinates()

        report_object = {
            'CID': self.cid,
            'ADOP': self.adop,
            'AMFR': self.amfr,
            'AMOD': self.amod,
            'APQS': self.apqs,
            'ASER': self.aser,
            'EDOP': self.edop,
            'EMFR': self.emfr,
            'EMOD': self.emod,
            'EPQS': self.epqs,
            'ESER': self.eser,
            'EMSV': self.emsv,
            'LDOP': self.ldop,
            'LMFR': self.lmfr,
            'LMOD': self.lmod,
            'LPQS': self.lpqs,
            'LSER': self.lser,
            'LSV' : self.lsv,
            'LAT': nudged_coordinates.latitude,
            'LNG': nudged_coordinates.longitude
        }

        if self.config.type == 'rtmd':
            report_object['AMID'] = self.amid
            report_object['DLST'] = self._build_dlst()
            report_object['records'] = recordset.to_rtmd()
            report = RtmdReport(**report_object)
        elif self.config.type == 'ems':
            report_object['records'] = recordset.to_ems(powersource=self.powersource)
            report = EmsReport(**report_object)

        self.last_sample_time = self.last_sample_time + dt.timedelta(seconds=batch_size * self.config.sample_interval)

        return report

File: js/src/device.js
/**
 * Device module — slimmed-down port of ccesim/device.py.
 *
 * No facility/device catalogs: the game provides all metadata via config.
 */

import { SimulatedRecordSet, SimulatorState } from "./recordset.js";
import { SimulationConfig, defaultConfig } from "./config.js";
import {
  RtmdReport,
  EmsReport,
  SCHEMA_VERSION,
  TRANSFER_SRC,
} from "./schemas.js";
import { SeededRandom } from "./random.js";

/**
 * Per-report coordinate jitter, in degrees. Mirrors
 * ccesim/facilities.py get_nudged_coordinates(), which draws
 * random.gauss(0.00001, 0.001) per axis.
 */
const COORD_JITTER_MU = 0.00001;
const COORD_JITTER_SIGMA = 0.001;

// ---------------------------------------------------------------------------
// Generator utilities (from ccesim/generator.py)
// ---------------------------------------------------------------------------

/**
 * Generate a random hex serial (UUID v4 without dashes).
 * @returns {string} 32-character hex string
 */
export function randomSerial() {
  return crypto.randomUUID().replace(/-/g, "");
}

/**
 * Generate a supplier-internal Appliance Monitoring ID (AMID).
 *
 * Per cce-interop, AMID references the appliance in the RTMD supplier's cloud
 * platform -- deliberately NOT the serial/asset number. A random 12-hex token
 * (e.g. "4bb74045097e") matches the schema examples; mint once per device and
 * reuse across reports to keep it stable.
 *
 * @returns {string} 12-character hex string
 */
export function randomAmid() {
  return crypto.randomUUID().replace(/-/g, "").slice(0, 12);
}

/**
 * Build a TransferMetadata plain object.
 * @param {'rtmd'|'ems'} type
 * @param {string|null} [callbackUrl] - Optional webhook URL. Per cce-interop the
 *   field is 'transferCallbackUrl' (non-nullable string); it is omitted entirely
 *   when no URL is supplied.
 * @param {string|null} [schemaVersion] - Override the declared cce-interop
 *   version; null keeps SCHEMA_VERSION. Relabels the transmission only -- the
 *   records are generated the same way either way.
 * @param {string|null} [transferSrc] - Override the data transmission source
 *   (the URI identifying the data supplier the delivery is attributed to);
 *   null keeps TRANSFER_SRC. Re-attributes the transmission only -- the
 *   records are unchanged.
 * @returns {object}
 */
export function transferMetadata(
  type = "rtmd",
  callbackUrl = null,
  schemaVersion = null,
  transferSrc = null,
) {
  const obj = {
    transferId: crypto.randomUUID(),
    transferSrc: transferSrc || TRANSFER_SRC,
    transferredAt: new Date(),
    transferType: type === "ems" ? "ems" : "rtm",
    schemaVersion: schemaVersion || SCHEMA_VERSION,
  };
  if (callbackUrl != null) {
    obj.transferCallbackUrl = callbackUrl;
  }
  return obj;
}

// ---------------------------------------------------------------------------
// MonitoringDeviceConfig
// ---------------------------------------------------------------------------

/**
 * Configuration for a virtual CCE monitoring device.
 *
 * All metadata is passed in explicitly — no random selection from catalogs.
 */
export class MonitoringDeviceConfig {
  /**
   * @param {object} opts
   * @param {'rtmd'|'ems'} opts.type - Device type.
   * @param {number} opts.uploadInterval - Seconds between uploads.
   * @param {number} opts.sampleInterval - Seconds between samples.
   * @param {'mains'|'solar'} [opts.powerType='mains'] - Power source.
   *
   * Appliance identity:
   * @param {string} opts.amfr - Appliance manufacturer.
   * @param {string} opts.amod - Appliance model.
   * @param {string} [opts.apqs] - Appliance PQS code.
   * @param {string} [opts.aser] - Appliance serial (auto-generated if omitted).
   * @param {string} [opts.adop] - Appliance date of placement (YYYY-MM-DD).
   *
   * Logger identity (RTMD only — ignored for EMS):
   * @param {string} [opts.lmfr] - Logger manufacturer.
   * @param {string} [opts.lmod] - Logger model.
   * @param {string} [opts.lpqs] - Logger PQS code.
   * @param {string} [opts.lser] - Logger serial (auto-generated if omitted).
   * @param {string} [opts.ldop] - Logger date of placement.
   *
   * Facility:
   * @param {string} [opts.cid] - Country ISO code.
   * @param {number} [opts.lat] - Latitude.
   * @param {number} [opts.lng] - Longitude.
   *
   * @param {SimulationConfig} [opts.simConfig] - Override simulation config.
   */
  constructor({
    type,
    uploadInterval,
    sampleInterval,
    powerType = "mains",

    amfr,
    amod,
    apqs = null,
    aser = null,
    adop = null,

    lmfr = null,
    lmod = null,
    lpqs = null,
    lser = null,
    ldop = null,

    cid = null,
    lat = null,
    lng = null,

    simConfig = null,
  }) {
    if (!type || !["rtmd", "ems"].includes(type)) {
      throw new Error(`type must be 'rtmd' or 'ems', got '${type}'`);
    }
    if (uploadInterval < sampleInterval) {
      throw new Error("uploadInterval must be >= sampleInterval");
    }
    if (uploadInterval % sampleInterval !== 0) {
      throw new Error("uploadInterval must be a multiple of sampleInterval");
    }

    this.type = type;
    this.uploadInterval = uploadInterval;
    this.sampleInterval = sampleInterval;
    this.powerType = powerType;
    this.batchSize = uploadInterval / sampleInterval;

    // Appliance
    this.amfr = amfr;
    this.amod = amod;
    this.apqs = apqs;
    this.aser = aser;
    this.adop = adop;

    // Logger (separate identity for RTMD; mirrors appliance for EMS)
    this.lmfr = lmfr;
    this.lmod = lmod;
    this.lpqs = lpqs;
    this.lser = lser;
    this.ldop = ldop;

    // Facility
    this.cid = cid;
    this.lat = lat;
    this.lng = lng;

    // Simulation config
    this.simConfig = simConfig;
  }
}

// ---------------------------------------------------------------------------
// BaseRtmDevice
// ---------------------------------------------------------------------------

/**
 * A simulated RTMD or EMS monitoring device.
 *
 * Call createReport(reportTime) repeatedly to produce sequential reports
 * with thermal state continuity.
 */
export class BaseRtmDevice {
  /**
   * @param {MonitoringDeviceConfig} config
   */
  constructor(config) {
    this.config = config;
    this.lastSampleTime = null;
    this.simulatorState = null;

    // Build simulation config
    this.simConfig =
      config.simConfig ??
      defaultConfig(config.powerType, config.lat);
    this.simConfig.sample_interval = config.sampleInterval;

    // Appliance identity
    this.amfr = config.amfr;
    this.amod = config.amod;
    this.apqs = config.apqs;
    this.aser = config.aser ?? randomSerial();
    this.adop = config.adop ?? _mockAdop();

    // Logger / EMD identity
    if (config.type === "rtmd") {
      this.lmfr = config.lmfr;
      this.lmod = config.lmod;
      this.lpqs = config.lpqs;
    } else {
      // EMS: logger = appliance
      this.lmfr = this.amfr;
      this.lmod = this.amod;
      this.lpqs = this.apqs;
    }
    this.lser = config.lser ?? randomSerial();
    this.ldop = config.ldop ?? this.adop;
    this.lsv = "0.1.x";

    // EMD mirrors logger
    this.emfr = this.lmfr;
    this.emod = this.lmod;
    this.epqs = this.lpqs;
    this.eser = this.lser;
    this.edop = this.ldop;
    this.emsv = this.lsv;

    // Facility
    this.cid = config.cid;
    this.lat = config.lat;
    this.lng = config.lng;

    // Supplier-internal appliance monitoring id, minted once and held stable
    // across all reports this device produces.
    this.amid = randomAmid();

    // RNG for the per-report coordinate jitter (see _nudgedCoordinates).
    // Deliberately its OWN stream: the record stream is owned by
    // SimulatedRecordSet.generate and is carried across reports in
    // simulatorState.rng_state, so drawing coordinates from it would shift
    // every measurement and break the cross-validation fixtures. Seeded from
    // the simulation config so a seeded device stays fully deterministic.
    this._coordRng = new SeededRandom(
      this.simConfig.random_seed ?? Date.now(),
    );
  }

  /**
   * The configured facility coordinates with a small random offset applied.
   * THIS IS THE ONLY COORDINATE PAIR THAT SHOULD REACH A REPORT.
   *
   * THE JITTER IS DELIBERATE -- do not remove it as noise. It mirrors the
   * Python implementation's Facility.get_nudged_coordinates(): a
   * gauss(0.00001, 0.001) degree offset per axis, about 111 m at one sigma,
   * so the exact point of a facility that may be a real one is blurred. It
   * does NOT anonymise it -- 111 m is still inside the compound. Treat it as
   * courtesy, never as a privacy control.
   *
   * It is drawn afresh on every call, so successive reports from one device
   * carry slightly different LAT/LNG, as a real GPS fix would. It never
   * mutates this.lat / this.lng, so anything reading the device's configured
   * position is unaffected.
   *
   * A coordinate that is not a finite number (the default is null) is passed
   * through untouched, so an unset LAT/LNG stays unset rather than becoming a
   * spurious point near the equator.
   *
   * @returns {{lat: number|null, lng: number|null}}
   */
  _nudgedCoordinates() {
    const nudge = (v) =>
      typeof v === "number" && Number.isFinite(v)
        ? v + this._coordRng.gauss(COORD_JITTER_MU, COORD_JITTER_SIGMA)
        : v;
    return { lat: nudge(this.lat), lng: nudge(this.lng) };
  }

  /**
   * Build the cce-interop DLST: maps the performance properties this RTMD
   * reports (TVC, TAMB) to sensor definitions. The sensor maker mirrors the
   * EMD/logger manufacturer (EMFR); SIDs are the logger serial plus a sensor
   * port, per the schema's $comment.
   *
   * @returns {object}
   */
  _buildDlst() {
    return {
      TVC: {
        SID: `${this.eser}-P1`,
        SMFR: this.emfr,
        SMOD: `${this.emod} External Probe`,
        SDOP: this.edop,
      },
      TAMB: {
        SID: `${this.eser}-P2`,
        SMFR: this.emfr,
        SMOD: `${this.emod} Onboard Sensor`,
        SDOP: this.edop,
      },
    };
  }

  /**
   * Create a report covering the period since the last report.
   *
   * @param {Date} reportTime - End of the reporting window (defaults to now).
   * @returns {RtmdReport|EmsReport}
   */
  createReport(reportTime = null) {
    if (reportTime === null) {
      reportTime = new Date();
    }
    if (!(reportTime instanceof Date)) {
      throw new Error("reportTime must be a Date");
    }

    if (this.lastSampleTime === null) {
      this.lastSampleTime = new Date(
        reportTime.getTime() - this.config.uploadInterval * 1000,
      );
    }

    if (this.lastSampleTime >= reportTime) {
      throw new Error(
        "reportTime must be after the last sample time",
      );
    }

    const gapMs = reportTime.getTime() - this.lastSampleTime.getTime();
    const batchSize = Math.floor(gapMs / (this.config.sampleInterval * 1000));
    const firstRecordTime = new Date(
      this.lastSampleTime.getTime() + this.config.sampleInterval * 1000,
    );

    // Generate simulated records
    const recordset = SimulatedRecordSet.generate(
      this.simConfig,
      batchSize,
      firstRecordTime,
      this.config.sampleInterval,
      this.simulatorState,
    );
    this.simulatorState = recordset.state;

    const nudgedCoordinates = this._nudgedCoordinates();

    const reportObj = {
      CID: this.cid,
      ADOP: this.adop,
      AMFR: this.amfr,
      AMOD: this.amod,
      APQS: this.apqs,
      ASER: this.aser,
      EDOP: this.edop,
      EMFR: this.emfr,
      EMOD: this.emod,
      EPQS: this.epqs,
      ESER: this.eser,
      EMSV: this.emsv,
      LDOP: this.ldop,
      LMFR: this.lmfr,
      LMOD: this.lmod,
      LPQS: this.lpqs,
      LSER: this.lser,
      LSV: this.lsv,
      LAT: nudgedCoordinates.lat,
      LNG: nudgedCoordinates.lng,
    };

    let report;
    if (this.config.type === "rtmd") {
      reportObj.AMID = this.amid;
      reportObj.DLST = this._buildDlst();
      reportObj.records = recordset.toRtmd();
      report = new RtmdReport(reportObj);
    } else {
      reportObj.records = recordset.toEms(this.config.powerType);
      report = new EmsReport(reportObj);
    }

    // Advance last sample time
    this.lastSampleTime = new Date(
      this.lastSampleTime.getTime() + batchSize * this.config.sampleInterval * 1000,
    );

    return report;
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Generate a random YYYY-MM-DD date between 2018-01-01 and 2024-12-31.
 * @returns {string}
 */
function _mockAdop() {
  const start = new Date("2018-01-01").getTime();
  const end = new Date("2024-12-31").getTime();
  const d = new Date(start + Math.random() * (end - start));
  return d.toISOString().slice(0, 10);
}
